### PR TITLE
Tokenizer summary

### PR DESCRIPTION
**Pull request by sgugger**
This PR introduces a mid/high-level summary of the different tokenizer types used in the library (a bit like the model summary).

Preview is [here](https://56179-155220641-gh.circle-artifacts.com/0/docs/_build/html/tokenizer_summary.html).

### CODE REVIEW

File: docs/source/tokenizer_summary.rst [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+will be different than the tokens "Transformers" and "do" for our model, so we should probably take the punctuation
+into account. This would give:
+
+::
+
+    ["Don", "'", "t", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]
+
+which is better already. One thing that is annoying though is how it dealt with "Don't". "Don't" stands for do not, so
+it should probably be better tokenized as ``["Do", "n't"]``. This is where things start getting more complicated, and
+part of the reason each kind of model as its own tokenizer class. Depending on the rules we apply to split our texts
```
Comment (n1t0) [inline docs/source/tokenizer_summary.rst:36]:
```suggestion
part of the reason each kind of model has its own tokenizer class. Depending on the rules we apply to split our texts
```

File: docs/source/tokenizer_summary.rst [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+::
+
+    ["Do", "n't", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]
+
+Space/punctuation-tokenization and rule-based tokenization are both examples of word tokenization, which is splitting a
+sentence into words. While it's the most intuitive way to separate texts in smaller chunks, it can have a problem when
+you have a huge corpus: it usually yields a very big vocabulary. :doc:`Transformer XL <model_doc/transformerxl>` for
+instance uses space/punctuation-tokenization, and has a vocabulary size of 267,735!
+
+A huge vocabulary size means a huge embedding matrix, which will cause memory problem. TransformerXL deals with it by
```
Comment (n1t0) [inline docs/source/tokenizer_summary.rst:52]:
```suggestion
A huge vocabulary size means a huge embedding matrix, which will cause memory problems. TransformerXL deals with it by
```

File: docs/source/tokenizer_summary.rst [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+Subword tokenization
+^^^^^^^^^^^^^^^^^^^^
+
+Subword tokenization algorithms rely on the principle that most common words should be left as is, but rare words
+should be decomposed in meaningful subword units. For instance "annoyingly" might be considered a rare word and
+decomposed as "annoying" and "ly". This is especially useful in agglutinative languages such as Turkish, where you can
+form (almost) arbitrarily long complex words by stringing together some subwords.
+
+This allows the model to keep a reasonable vocabulary while still learning useful representations for common words or
+subwords. This also give the ability to the model to process words it has never seen before, by decomposing them into
```
Comment (n1t0) [inline docs/source/tokenizer_summary.rst:70]:
```suggestion
subwords. This also gives the ability to the model to process words it has never seen before, by decomposing them into
```

File: docs/source/tokenizer_summary.rst [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+Byte-Pair Encoding
+~~~~~~~~~~~~~~~~~~
+
+Byte-Pair Encoding was introduced in `this paper <https://arxiv.org/abs/1508.07909>`__. It relies on a pretokenizer
+splitting the training data into words, which can be a simple space tokenization
+(:doc:`GPT-2 <model_doc/gpt2>` and :doc:`Roberta <model_doc/roberta>` uses this for instance) or a rule-based tokenizer
+(:doc:`XLM <model_doc/xlm>` use Moses for most languages, as do :doc:`FlauBERT <model_doc/flaubert>`),
+:doc:`GPT <model_doc/gpt>` uses Spacy and ftfy) and, counts the frequency of each word in the training corpus.
+
+It then begins from the list of all characters, and will learn merge rules to form a new token from to symbols in the
```
Comment (n1t0) [inline docs/source/tokenizer_summary.rst:112]:
```suggestion
It then begins from the list of all characters, and will learn merge rules to form a new token from two symbols in the
```

File: docs/source/tokenizer_summary.rst [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+to the vocabulary.
+
+At this stage, the vocabulary is ``['b', 'g', 'h', 'n', 'p', 's', 'u', 'ug', 'un', 'hug']`` and our corpus is
+represented as
+
+::
+
+    ('hug', 10), ('p' 'ug', 5), ('p' 'un', 12), ('b' 'un', 4), ('hug' 's', 5)
+
+If we stop there, the tokenizer can apply the rules it learned to new words (as long as they don't have characters that
```
Comment (n1t0) [inline docs/source/tokenizer_summary.rst:148]:
```suggestion
If we stop there, the tokenizer can apply the rules it learned to new words (as long as they don't contain characters that
```

File: docs/source/tokenizer_summary.rst [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+are in the base vocabulary). For instance 'bug' would be tokenized as ``['b', 'ug']`` but mug would be tokenized as
```
Comment (n1t0) [inline docs/source/tokenizer_summary.rst:149]:
```suggestion
were not in the base vocabulary). For instance 'bug' would be tokenized as ``['b', 'ug']`` but mug would be tokenized as
```

File: docs/source/tokenizer_summary.rst [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+``['<unk>', 'ug']`` since the 'm' is not in the base vocabulary. This doesn't happen to letters in general (since the
+base corpus uses all of them), but to special characters like emoticons.
```
Comment (n1t0) [inline docs/source/tokenizer_summary.rst:151]:
```suggestion
base corpus uses all of them), but to special characters like emojis.
```

**Review by n1t0 — APPROVED** [review]
Looks good to me, very well-explained @sgugger! I just added a few suggestions for small typos.

File: docs/source/tokenizer_summary.rst [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+rule-based tokenizers. On the text above, they'd output something like:
+
+::
+
+    ["Do", "n't", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]
+
+Space/punctuation-tokenization and rule-based tokenization are both examples of word tokenization, which is splitting a
+sentence into words. While it's the most intuitive way to separate texts in smaller chunks, it can have a problem when
+you have a huge corpus: it usually yields a very big vocabulary. :doc:`Transformer XL <model_doc/transformerxl>` for
+instance uses space/punctuation-tokenization, and has a vocabulary size of 267,735!
```
Comment (LysandreJik) [inline docs/source/tokenizer_summary.rst:50]: In these lines as well as the following you often mention the word "vocabulary". As a newcomer, I'm not sure I would understand what that is.

File: docs/source/tokenizer_summary.rst (hunk 3 above)
Comment (LysandreJik) [inline docs/source/tokenizer_summary.rst:70]:
```suggestion
subwords. This also gives the ability to the model to process words it has never seen before, by decomposing them into

```

File: docs/source/tokenizer_summary.rst [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+.. _byte-pair-encoding:
+
+Byte-Pair Encoding
+~~~~~~~~~~~~~~~~~~
+
+Byte-Pair Encoding was introduced in `this paper <https://arxiv.org/abs/1508.07909>`__. It relies on a pretokenizer
+splitting the training data into words, which can be a simple space tokenization
+(:doc:`GPT-2 <model_doc/gpt2>` and :doc:`Roberta <model_doc/roberta>` uses this for instance) or a rule-based tokenizer
+(:doc:`XLM <model_doc/xlm>` use Moses for most languages, as do :doc:`FlauBERT <model_doc/flaubert>`),
```
Comment (LysandreJik) [inline docs/source/tokenizer_summary.rst:109]:
```suggestion
(:doc:`XLM <model_doc/xlm>` use Moses for most languages, as does :doc:`FlauBERT <model_doc/flaubert>`),

```

File: docs/source/tokenizer_summary.rst (hunk 4 above)
Comment (LysandreJik) [inline docs/source/tokenizer_summary.rst:112]:
```suggestion
It then begins from the list of all characters, and will learn merge rules to form a new token from two symbols in the

```

File: docs/source/tokenizer_summary.rst (hunk 7 above)
Comment (LysandreJik) [inline docs/source/tokenizer_summary.rst:151]:
```suggestion
base corpus uses all of them), but to special characters like emojis.

```

**Review by LysandreJik — APPROVED** [review]
Very cool!

File: docs/source/tokenizer_summary.rst (hunk 8 above)
Comment (sgugger) [reply docs/source/tokenizer_summary.rst:50]: Good point, will try to add some definition.

**Pull request merged**